content/300-guides/050-database/900-turso.mdx

```diff
@@ -0,0 +1,183 @@
+
+To update your database schema:
+
+1. Generate a migration file using `prisma migrate dev` against a local SQLite database:
+
+```sh
+npx prisma migrate dev --name init # Migration name
+```
+
+2. Apply the migration using Turso's CLI:
+
+```sh
+turso db shell turso-prisma-db < ./prisma/migrations/20230922132717_init/migration.sql # Replace `20230922132717_init` with the existing migration
+```
+
+This workflow does not support tracking the history of applied migrations.
+
+## How to connect and query from an embedded replica
+
+Turso supports [embedded replicas](https://blog.turso.tech/introducing-embedded-replicas-deploy-turso-anywhere-2085aa0dc242). Turso's embedded replicas enable you to have a copy of your primary, remote database _inside_ your application. This is similar to having a local SQLite database.
+
+When your app initially establishes a connection to your database, the primary database will fulfill the query:
+
+![Embedded Replica: First remote read](./images/embedded-replica-remote-read.png)
+
+Turso will (1) create an embedded replica inside your application and (2) copy data from your primary database to the replica so it is locally available:
+
+![Embedded Replica: Remote DB Copy](./images/embedded-replica-create-replica.png)
+
+The embedded replica will fulfill subsequent read queries. The libSQL client provides a [`sync()`](<https://docs.turso.tech/libsql/client-access/javascript-typescript-sdk#client-capability-summary:~:text=an%20interactive%20transaction-,sync(),-Synchronize%20the%20embedded>) method which you can invoke to ensure the embedded replica's data remains fresh.
+
+![Embedded Replica: Local DB reads](./images/embedded-replica-read.png)
+
+With embedded replicas, this setup guarantees a responsive application, because the data will be readily available locally and faster to access.
+
+Like a read replica setup you may be familiar with, write operations are forwarded to the primary remote database and executed before being propagated to all embedded replicas.
+
+![Embedded Replica: Write operation propagation](./images/embedded-replica-write-propagation.png)
+
+1. Write operations propagation are forwarded to the database.
+1. Database responds to the server with the updates from 1.
+1. Write operations are propagated to the database replica.
+
+To get started using embedded replicas with Prisma, update your Prisma Client instance with the following changes:
+
+```ts highlight=6,7,11,16;add
+import { PrismaClient } from '@prisma/client'
+import { PrismaLibSQL } from '@prisma/adapter-libsql'
+import { createClient } from '@libsql/client'
+
+const libsql = createClient({
+  url: 'file:replica.db',
+  syncUrl: `${process.env.TURSO_DATABASE_URL}`,
+  authToken: `${process.env.TURSO_AUTH_TOKEN}`,
+})
+
+const sync = async () => await libsql.sync()
+
+const adapter = new PrismaLibSQL(libsql)
+
+const prismaClientSingleton = () => {
+  await sync()
+  return new PrismaClient({ adapter })
+}
+```
```

Comment:
Do we have an example for this that works and is tested that we can link to (full sourcecode)?

Reply:
We don't have an official repository (the one I built for the blog post), but I can whip up a minimal API example for working with embedded replicas.